File: tags/last-successful/jWebSocket/jWebSocket-1.0-YYYY-MM-DD-HH-MM-SS/shared/J2SE/jWebSocketCommon/org/jwebsocket/api/WebSocketPacket.java
//	---------------------------------------------------------------------------
//	jWebSocket - WebSocketPacket (Community Edition, CE)
//	---------------------------------------------------------------------------
//	Copyright 2010-2014 Innotrade GmbH (jWebSocket.org)
//	Alexander Schulze, Germany (NRW)
//
//	Licensed under the Apache License, Version 2.0 (the "License");
//	you may not use this file except in compliance with the License.
//	You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
//	Unless required by applicable law or agreed to in writing, software
//	distributed under the License is distributed on an "AS IS" BASIS,
//	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//	See the License for the specific language governing permissions and
//	limitations under the License.
//	---------------------------------------------------------------------------
package org.jwebsocket.api;

import java.io.UnsupportedEncodingException;
import java.util.Date;
import org.jwebsocket.kit.WebSocketFrameType;

/**
 * Specifies the API for low level data packets which are interchanged between
 * client and server. Data packets do not have a special format at this
 * communication level.
 *
 * @author Alexander Schulze
 */
public interface WebSocketPacket {

	/**
	 * Sets the value of the data packet to the given array of bytes.
	 *
	 * @param aByteArray
	 */
	void setByteArray(byte[] aByteArray);

	/**
	 *
	 * @return
	 */
	boolean isFragment();

	/**
	 *
	 * @param aDate
	 */
	void setCreationDate(Date aDate);

	/**
	 *
	 * @return
	 */
	Date getCreationDate();

	/**
	 * Sets the value of the data packet to the given string by using default
	 * encoding.
	 *
	 * @param aString
	 */
	void setString(String aString);

	/**
	 * Sets the value of the data packet to the given string by using the passed
	 * encoding.
	 *
	 * @param aString
	 * @param aEncoding
	 * @throws UnsupportedEncodingException
	 */
	void setString(String aString, String aEncoding) throws UnsupportedEncodingException;

	/**
	 * Sets the value of the data packet to the given string by using UTF-8
	 * encoding.
	 *
	 * @param aString
	 */
	void setUTF8(String aString);

	/**
	 * Sets the value of the data packet to the given string by using 7 bit
	 * US-ASCII encoding.
	 *
	 * @param aString
	 */
	void setASCII(String aString);

	/**
	 * Returns the content of the data packet as an array of bytes.
	 *
	 * @return Data packet as array of bytes.
	 */
	byte[] getByteArray();

	/**
	 * Returns the content of the data packet as a string using default
	 * encoding.
	 *
	 * @return Raw Data packet as string with default encoding.
	 */
	String getString();

	/**
	 * Returns the content of the data packet as a string using the passed
	 * encoding.
	 *
	 * @param aEncoding
	 * @return String using the passed encoding.
	 * @throws UnsupportedEncodingException
	 */
	String getString(String aEncoding) throws UnsupportedEncodingException;

	/**
	 * Interprets the data packet as a UTF8 string and returns the string in
	 * UTF-8 encoding. If an exception occurs "null" is returned.
	 *
	 * @return Data packet as UTF-8 string or {@code null} if not convertible.
	 */
	String getUTF8();

	/**
	 * Interprets the data packet as a US-ASCII string and returns the string in
	 * US-ASCII encoding. If an exception occurs "null" is returned.
	 *
	 * @return Data packet as US-ASCII string or {@code null} if not
	 * convertible.
	 */
	String getASCII();

	/**
	 *
	 * @return
	 */
	WebSocketFrameType getFrameType();

	/**
	 *
	 * @param aFrameType
	 */
	void setFrameType(WebSocketFrameType aFrameType);

	/**
	 *
	 * @return
	 */
	Integer size();
}

File: tags/last-successful/jWebSocket/jWebSocket-1.0-2015-03-03_22-55-24/jWebSocketCSClient/SourceCode/WebSocket/WebSocket/org/jwebsocket/protocol/WebSocketPacket.cs
/*--------------------------------------------------------------------------------
 * jWebSocket - WebSocketPacket
 * Copyright (c) 2013 Rolando Betancourt Toucet
 * -------------------------------------------------------------------------------
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 3 of the License, or (at your
 * option) any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for
 * more details.
 * You should have received a copy of the GNU Lesser General Public License along
 * with this program; if not, see <http://www.gnu.org/licenses/lgpl.html>.
 * -------------------------------------------------------------------------------
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using WebSocket.org.jwebsocket.protocol.api;
using WebSocket.org.jwebsocket.protocol.kit;
using WebSocket.org.jwebsocket.common;

namespace WebSocket.org.jwebsocket.protocol
{

    /// <author>Rolando Betancourt Toucet</author>
    /// <lastUpdate>4/13/2013</lastUpdate>
    /// <summary>
    /// Implements the low level data packets which are interchanged between
    /// client and server. Data packets do not have a special format at this
    /// communication level.
    /// </summary>
    public sealed class WebSocketPacket : IWebSocketPacket 
    {
        private byte[] mByteArray = null;
        private string[] mFragments = null;

        private int mFragmentsLoaded = 0;
        private int mFragmentsExpected = 0;

        private bool mIsFragmented = false;
        private bool mIsComplete = false;

        private DateTime mCreationDate = DateTime.Now;
        private long mTimeout = 0;
        private WebSocketFrameType mFrameType = WebSocketFrameType.TEXT;

        public WebSocketPacket(int aInitialSize)
        {
            InitFragmented(aInitialSize);
        }

        public WebSocketPacket(byte[] aByteArray)
        {
            mByteArray = aByteArray;
        }

        public WebSocketPacket(WebSocketFrameType aFrameType, byte[] aByteArray)
        {
            mFrameType = aFrameType;
            mByteArray = aByteArray;
        }

        public WebSocketPacket(string aString)
        {
            SetString(aString);
        }

        public WebSocketPacket(WebSocketFrameType aFrameType, string aString)
        {
            mFrameType = aFrameType;
            SetString(aString);
        }

        public WebSocketPacket(string aString, WebSocketTypeEncoding aEncoding)
        {
            SetString(aString, aEncoding);
        }

        public void InitFragmented(int aTotal)
        {
            mFragmentsExpected = aTotal;
            mFragments = new string[aTotal];
        }

        public void SetFragment(string aString, int aIdx)
        {
            mFragments[aIdx] = aString;
            mFragmentsLoaded++;
            mIsComplete = mFragmentsLoaded >= mFragmentsExpected;
        }

        public void PackFragments()
        {
            StringBuilder lSB = new StringBuilder();
            for (int i = 0; i < mFragments.Length; i++)
            {
                lSB.Append(mFragments[i]);
                mFragments[i] = null;
            }
            mByteArray = WebSocketConvert.StringToBytes(lSB.ToString(), WebSocketTypeEncoding.UTF8);
        }

        public void SetString(string aString)
        {
            mByteArray = WebSocketConvert.StringToBytes(aString, WebSocketTypeEncoding.Default);
        }

        public void SetString(string aString, WebSocketTypeEncoding aEncoding)
        {
            mByteArray = WebSocketConvert.StringToBytes(aString, aEncoding);
        }

        public void SetUTF8(string aString)
        {
            mByteArray = WebSocketConvert.StringToBytes(aString, WebSocketTypeEncoding.UTF8);
        }

        public void SetASCII(string aString)
        {
            mByteArray = WebSocketConvert.StringToBytes(aString, WebSocketTypeEncoding.ASCII);
        }

        public string GetString()
        {
            return WebSocketConvert.BytesToString(mByteArray, WebSocketTypeEncoding.Default);
        }

        public string GetString(WebSocketTypeEncoding aEncoding)
        {
            return WebSocketConvert.BytesToString(mByteArray, aEncoding);
        }

        public string GetUTF8()
        {
            try
            {
                return WebSocketConvert.BytesToString(mByteArray, WebSocketTypeEncoding.UTF8);
            }
            catch (Exception)
            {
                return null;
            }
        }

        public string GetASCII()
        {
            try
            {
                return WebSocketConvert.BytesToString(mByteArray, WebSocketTypeEncoding.ASCII);
            }
            catch (Exception)
            {
                return null;
            }
        }

        public WebSocketFrameType FrameType
        {
            get { return mFrameType; }
            set { mFrameType = value; }
        }

        public byte[] ByteArray
        {
            get { return mByteArray; }
            set { mByteArray = value; }
        }

        public string[] Fragments
        {
            get { return mFragments; }
            set { mFragments = value; }
        }

        public bool IsFragmented
        {
            get { return mIsFragmented; }
            set { mIsFragmented = value; }
        }

        public bool IsComplete
        {
            get { return mIsComplete; }
        }

        public DateTime CreationDate
        {
            get { return mCreationDate; }
            set { mCreationDate = value; }
        }

        public long Timeout
        {
            get { return mTimeout; }
            set { mTimeout = value; }
        }

    }
}
